File: scripts/verify/facade-parity.mjs
/**
 * facade-parity.mjs — does the Python port elect the same fourteen tones the
 * browser does, and give every building the same bucket?
 *
 * WHAT IS BEING PROVED. `scripts/bake_facades.py` is a transcription of
 * `window.quantiseFacades`. A transcription checked by re-reading it proves
 * nothing: the same misunderstanding writes both sides. So this runs the REAL
 * browser functions on the REAL data and dumps what they decided.
 *
 * IT RUNS THE WHOLE ASSEMBLY, not just the election, because the election is
 * over whatever the assembly produced:
 *
 *   mergeCapitolScene()  patches 12 buildings from capitol_overrides.json,
 *                        appends 604 from capitol.geojson, and registers
 *                        FACADE_PROTECTED
 *   applyUnion24()       rewrites one building's height AND its wd/wg/wn
 *   quantiseFacades()    elects fourteen tones over all of that
 *
 * Getting the assembly wrong and the algorithm right still gives the wrong
 * answer, and it would be an easy thing to miss — which is exactly why this
 * calls the real ones rather than reproducing them here.
 *
 * ON ITS OWN COPY, like outer-facade-parity.mjs. Fetching the documents again
 * keeps the check independent of what the live scene happens to be doing, and
 * `quantiseFacades` recomputes from `wd` every time, so running it twice on
 * separate data is safe.
 *
 * ONE KNOWN DUPLICATE, and it is harmless. `mergeCapitolScene` appends to the
 * global `window.FACADE_PROTECTED`, which the page's own load already filled,
 * so calling it again doubles that list. `quantiseFacades` dedupes protected
 * entries by their coarse key (`if (protectedKeys.has(key)) continue`), so the
 * second copy is dropped. Noted rather than worked around, because working
 * around it would mean not calling the real function.
 *
 * THE JOIN KEY IS `id`, which every snapshot building carries. Features without
 * one — the Capitol Complex is authored, not Overture — are compared by their
 * position in the assembled array instead, and the two sides build that array
 * the same way by construction.
 *
 * IT DUMPS THE ELECTION'S INPUTS AS WELL AS ITS OUTPUT, and that is not
 * padding. `wd`/`wg`/`wn` are what the election reads and
 * `building_class`/`final_height` are what `familyFor` reads, so with them in
 * the capture the comparator can say WHICH of the two sides is wrong —
 * assembly or algorithm — without a second browser run. A browser run here is
 * ~40 s and a wrong diagnosis costs an hour, so the ~500 KB is cheap. It is
 * also the only way to catch the assembly agreeing on COUNTS while disagreeing
 * on VALUES, which a `12 / 604 / 1` line cannot see.
 *
 * IT LOADS THE PAGE TWICE, and the second load answers a different question.
 * Pass A forces the browser to ELECT (`?bakedfacades=0`) and is what the Python
 * side is compared against — that proves the PORT. Pass B arms the baked
 * palette and is diffed against pass A here — that proves the SWITCH, which is
 * what the site actually runs. Both guards in js/facades.js fall back to
 * electing, and a fallback makes that diff come out perfect, so pass B asserts
 * `facadePaletteSource()` really says `baked` before the diff is believed.
 *
 * Writes scripts/verify/out/facade-browser.json.
 * Compare with:  python scripts/verify/facade_parity.py
 *
 * Usage: VERIFY_URL=http://127.0.0.1:8123 node scripts/verify/facade-parity.mjs
 */
import { chromium } from 'playwright-core';
import { BASE, launch } from './chrome.mjs';
import fs from 'node:fs';
import path from 'node:path';

// Resolved from THIS FILE, not from the working directory, and that is a fix
// rather than a tidy-up. `path.join('scripts','verify','out')` is relative to
// wherever node was launched, and this directory's README tells you to run its
// scripts FROM `scripts/verify` — so a run started the documented way wrote the
// capture to `scripts/verify/scripts/verify/out/`, silently, while
// `facade_parity.py` kept reading the real path and comparing against whatever
// capture happened to be sitting there. On 2026-08-27 that was a 23-day-old
// file from snapshot 2026-08-04: the harness reported 17 findings, 1,274
// mismatched families and "NOT a bijection", none of which were real, and the
// obvious reading of that output is that the change under test broke the bake.
// The stray nested copy had also been committed. A capture that lands in a
// different place depending on your shell's cwd is not a capture.
const OUT_DIR = path.join(path.dirname(new URL(import.meta.url).pathname.replace(/^\/([A-Za-z]:)/, '$1')), 'out');
fs.mkdirSync(OUT_DIR, { recursive: true });

const browser = await launch(chromium);

/**
 * One assembled-and-quantised run of the page.
 *
 * `wantBaked` is the ONE load-bearing argument. js/facades.js adopts
 * data/facade_palette.json when it is current, which is the whole point of the
 * port — and it turns the Python comparison circular the moment it is left on:
 * quantiseFacades would read back the file bake_facades.py wrote,
 * facade_parity.py would compare the bake against itself, and the run would
 * print a triumphant 3,057/3,057 while proving nothing at all. So pass A forces
 * the election with `bakedfacades=0` and that is what the Python side is held
 * to; pass B arms the bake and is diffed against pass A here. Each run asserts
 * which path it actually took, because a silent circular pass is a far worse
 * outcome than a failure.
 */
async function capture(wantBaked) {
  const page = await browser.newPage({ viewport: { width: 900, height: 600 } });
  page.on('pageerror', e => console.log('  [pageerror] ' + e.message));
  const flag = wantBaked ? '' : '&bakedfacades=0';
  await page.goto(BASE + '/index.html?intro=0&drift=0' + flag,
                  { waitUntil: 'networkidle', timeout: 120000 });
  await page.waitForFunction(() => window.__map && window.__map.isStyleLoaded(), null, { timeout: 120000 });
  await page.evaluate(() => window.cancelGraphicsAutoDetect && window.cancelGraphicsAutoDetect());
  await page.waitForTimeout(2500);
  const r = await page.evaluate(async (wantBaked) => {
  for (const fn of ['quantiseFacades', 'mergeCapitolScene', 'applyUnion24']) {
    if (typeof window[fn] !== 'function') return { error: fn + ' is not defined' };
  }
  if (window.FACADE_BAKED_ON !== !!wantBaked) {
    return { error: 'FACADE_BAKED_ON is ' + window.FACADE_BAKED_ON
                  + ' but this pass wanted ' + !!wantBaked + ' — the flag did not take.' };
  }
  // Let the module's own fetch land before quantising, so a slow disk cannot
  // make pass B quietly fall back to electing and then be reported as a match.
  if (window.facadeBakedReady) await window.facadeBakedReady();
  const manifest = await fetch('data/manifest.json').then(r => r.json()).catch(() => null);
  const date = (manifest && (manifest.latest || manifest.date)) ||
    (manifest && Array.isArray(manifest.snapshots) && manifest.snapshots[manifest.snapshots.length - 1]);
  if (!date) return { error: 'could not resolve the snapshot date from manifest.json' };

  const buildings = await fetch(`data/snapshots/${date}/buildings.detailed.geojson`).then(r => r.json());
  const parts = await fetch(`data/snapshots/${date}/parts.detailed.geojson`).then(r => r.json());

  const merged = await window.mergeCapitolScene(buildings, parts);
  const u24 = window.applyUnion24(buildings.features);
  const stats = window.quantiseFacades(buildings.features);

  return {
    date,
    stats,
    total: buildings.features.length,
    source: window.facadePaletteSource ? window.facadePaletteSource() : 'unknown',
    // Reported by the real functions, so the Python assembly counts are checked
    // against the browser's own claim rather than against a re-count.
    assembly: { patched: merged && merged.patched, capitol: merged && merged.added,
                union24: u24 ? 1 : 0 },
    protectedLen: Array.isArray(window.FACADE_PROTECTED) ? window.FACADE_PROTECTED.length : 0,
    // Palette is read back off the module rather than returned by the call —
    // quantiseFacades returns only counts. NOTE the accessor is
    // `window.facadePalette`, not `__facadePalette`: the first draft of this
    // file guessed the name and would have silently captured `null`, which the
    // comparator would have read as "the browser has no palette" instead of as
    // "this script asked the wrong question".
    palette: (typeof window.facadePalette === 'function') ? window.facadePalette() : null,
    rows: buildings.features.map((f, i) => {
      const p = f.properties || {};
      return { i, id: p.id != null ? String(p.id) : null,
               wd: p.wd || null, wg: p.wg || null, wn: p.wn || null,
               bc: p.building_class || null, fh: p.final_height == null ? null : p.final_height,
               wp: p.wp || null, wf: p.wf || null };
    }),
  };
  }, wantBaked);
  await page.close();
  return r;
}

let fail = 0;
const say = m => { console.log(m); if (m.startsWith('*FAIL')) fail = 1; };

// ── PASS A — the browser elects, and the Python side is held to this ────────
const A = await capture(false);
if (A.error) {
  say('*FAIL — pass A: ' + A.error);
} else {
  const stamped = A.rows.filter(r => r.wp).length;
  const buckets = new Set(A.rows.filter(r => r.wp).map(r => r.wp.slice(2)));
  console.log(`snapshot ${A.date}`);
  console.log(`assembled ${A.total} buildings, browser stamped ${stamped}`);
  console.log(`quantiseFacades reported ${JSON.stringify(A.stats)}`);
  console.log(`assembly ${JSON.stringify(A.assembly)}`);
  console.log(`palette source: ${A.source}`);
  console.log(`FACADE_PROTECTED carries ${A.protectedLen} entries (the page's own load + this one)`);
  console.log(`palette entries captured: ${A.palette ? A.palette.length : 'NONE'}`);
  console.log(`distinct bucket ordinals in use: ${buckets.size}`);
  const file = path.join(OUT_DIR, 'facade-browser.json');
  fs.writeFileSync(file, JSON.stringify(A));
  console.log('\nwrote ' + file + '  (' + (fs.statSync(file).size / 1024).toFixed(0) + ' KB)');
  if (stamped === 0) say('*FAIL — the browser stamped nothing');
  if (!A.palette || !A.palette.length) {
    say('*FAIL — no palette captured; window.facadePalette is missing or empty');
  }

  // ── PASS B — the baked palette armed, diffed against pass A ──────────────
  //
  // The Python comparison proves the PORT. This proves the SWITCH, which is a
  // different claim and the one the site actually runs: that a page which takes
  // its fourteen buckets from data/facade_palette.json paints exactly the city
  // a page that elects them does. It has to be a second page load rather than a
  // second call on this one, because the adoption is decided at script-parse
  // time from the URL.
  console.log('\n── pass B: the baked palette, armed ──');
  const B = await capture(true);
  if (B.error) {
    say('*FAIL — pass B: ' + B.error);
  } else {
    console.log(`palette source: ${B.source}`);
    console.log(`quantiseFacades reported ${JSON.stringify(B.stats)}`);
    if (!/^baked /.test(B.source)) {
      // The whole point. Both guards fall back to electing, and a fallback
      // produces an identical diff — so without this check pass B passes
      // loudest exactly when the baked palette was never used.
      say('*FAIL — pass B did not adopt the bake (' + B.source
          + '); the diff below would be the elected path against itself');
    }
    if (B.total !== A.total) {
      say(`*FAIL — feature count moved between passes: ${A.total} vs ${B.total}`);
    } else {
      let wpBad = 0, wfBad = 0, first = null;
      for (let i = 0; i < A.rows.length; i++) {
        if (A.rows[i].wp !== B.rows[i].wp) { wpBad++; first = first || [i, A.rows[i].wp, B.rows[i].wp]; }
        if (A.rows[i].wf !== B.rows[i].wf) wfBad++;
      }
      let palBad = 0;
      const ap = A.palette || [], bp = B.palette || [];
      if (ap.length !== bp.length) palBad++;
      for (let i = 0; i < Math.min(ap.length, bp.length); i++) {
        for (const k of ['wd', 'wg', 'wn']) if (ap[i][k] !== bp[i][k]) palBad++;
      }
      console.log(`wp identical      ${A.rows.length - wpBad} / ${A.rows.length}`);
      console.log(`wf identical      ${A.rows.length - wfBad} / ${A.rows.length}`);
      console.log(`palette entries   ${palBad ? palBad + ' DIFFER' : 'all ' + ap.length + ' identical'}`);
      if (wpBad || wfBad || palBad) {
        say(`*FAIL — elected and baked disagree (first: #${first && first[0]} `
            + `${first && first[1]} vs ${first && first[2]})`);
      }
    }
  }
}

console.log('');
if (fail) {
  process.exitCode = 1;
} else {
  console.log('PASS — the baked palette paints the same city the election does.');
  console.log('now run:  python scripts/verify/facade_parity.py');
}

await browser.__done();

File: scripts/verify/facade_parity.py
# -*- coding: utf-8 -*-
"""Compare the Python campus facade election against what the browser did.

Reads scripts/verify/out/facade-browser.json, written by
scripts/verify/facade-parity.mjs from a live run of the REAL
`mergeCapitolScene` / `applyUnion24` / `quantiseFacades`.

WHY THIS IS FOUR CHECKS AND NOT ONE. The outer-ring port could only compare
the PARTITION, because the browser names a tower bucket `tg<n>` where n is an
offset into the campus palette and the two sides cannot agree on the string by
construction. Here there is no offset: the campus palette IS the whole palette
and the ordinal is the palette index, so this can demand something much
stronger than a bijection -- the same ordinal, the same family, the same
fourteen hex triples, on the same feature. Anything weaker would pass a build
that renders every wall the wrong colour in a self-consistent way.

  1. ASSEMBLY COUNTS   12 patched / 604 appended / 1 Union24, and the total.
  2. ASSEMBLY VALUES   `wd` `wg` `wn` `building_class` `final_height` for all
                       3,057 features, positionally. This is the check the
                       counts cannot make: the two sides agreeing that 12
                       buildings were patched says nothing about whether they
                       were patched to the same colours, and every one of those
                       five fields is an INPUT to the election or to familyFor.
  3. PALETTE           the fourteen {wd,wg,wn} hex triples, in order, exact.
  4. STAMPS            `wp` and `wf` per feature, exact string equality, plus
                       the bijection both ways as a diagnostic so a failure
                       says whether the labelling is merely OFFSET or actually
                       DIFFERENT.

THE JOIN IS POSITIONAL, and it has to be. 604 of the 3,057 features are the
authored Capitol Complex and carry no `id`, so an id-keyed join would check
80% of the city and report a pass. Both sides build the array the same way --
snapshot order, then capitol.geojson appended -- and check 2 is what proves
that claim rather than assuming it.

Usage:  python scripts/verify/facade_parity.py
"""
import json
import os
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.dirname(os.path.dirname(HERE))
sys.path.insert(0, os.path.join(ROOT, "scripts"))

from bake_facades import load_scene, quantise  # noqa: E402

BROWSER = os.path.join(HERE, "out", "facade-browser.json")
MAX_SHOWN = 12


def _num(x):
    """`final_height` crosses JSON as a float on one side and may be an int on
    the other; compare as numbers, and treat null/0 the same way `|| 0` does."""
    return float(x or 0)


def main():
    if not os.path.exists(BROWSER):
        print("*FAIL - %s is missing. Run facade-parity.mjs first."
              % os.path.relpath(BROWSER, ROOT).replace("\\", "/"))
        return 1
    b = json.load(open(BROWSER, encoding="utf-8"))
    brows = b["rows"]

    date, feats, protected, asm = load_scene()
    palette, index, rows, combos = quantise(feats, protected)

    fails = []

    # ---------------------------------------------------------- 1. assembly --
    if date != b["date"]:
        fails.append("snapshot date: python %s vs browser %s" % (date, b["date"]))
    if len(rows) != len(brows):
        fails.append("feature count: python %d vs browser %d" % (len(rows), len(brows)))
    for k in ("patched", "capitol", "union24"):
        if asm.get(k) != b["assembly"].get(k):
            fails.append("assembly.%s: python %s vs browser %s"
                         % (k, asm.get(k), b["assembly"].get(k)))

    n = min(len(rows), len(brows))

    # ---------------------------------------------- 2. the election's INPUTS --
    field_bad = {}
    for i in range(n):
        p = feats[i].get("properties") or {}
        q = brows[i]
        pairs = (
            ("id", None if p.get("id") is None else str(p["id"]), q["id"]),
            ("wd", p.get("wd") or None, q["wd"]),
            ("wg", p.get("wg") or None, q["wg"]),
            ("wn", p.get("wn") or None, q["wn"]),
            ("building_class", p.get("building_class") or None, q["bc"]),
        )
        for name, mine, theirs in pairs:
            if mine != theirs:
                field_bad.setdefault(name, []).append((i, mine, theirs))
        if _num(p.get("final_height")) != _num(q["fh"]):
            field_bad.setdefault("final_height", []).append(
                (i, p.get("final_height"), q["fh"]))
    for name, bad in sorted(field_bad.items()):
        fails.append("assembly value %s differs on %d features, e.g. %s"
                     % (name, len(bad), "; ".join(
                         "#%d python=%r browser=%r" % t for t in bad[:3])))

    # ----------------------------------------------------------- 3. palette --
    bp = b.get("palette") or []
    if len(bp) != len(palette):
        fails.append("palette length: python %d vs browser %d" % (len(palette), len(bp)))
    for i in range(min(len(bp), len(palette))):
        for k in ("wd", "wg", "wn"):
            if palette[i][k].lower() != str(bp[i].get(k, "")).lower():
                fails.append("palette[%d].%s: python %s vs browser %s"
                             % (i, k, palette[i][k], bp[i].get(k)))

    # ------------------------------------------------------------ 4. stamps --
    wp_bad, wf_bad, stamped_bad = [], [], []
    fwd, rev = {}, {}
    for i in range(n):
        mine, theirs = rows[i], brows[i]
        if (mine is None) != (theirs["wp"] is None):
            stamped_bad.append((i, mine and mine["wp"], theirs["wp"]))
            continue
        if mine is None:
            continue
        if mine["wp"] != theirs["wp"]:
            wp_bad.append((i, mine["wp"], theirs["wp"]))
        if mine["wf"] != theirs["wf"]:
            wf_bad.append((i, mine["wf"], theirs["wf"]))
        # Bucket-ordinal bijection, both directions, as a DIAGNOSTIC. If the
        # exact check above fails, this says which kind of failure it is: a
        # clean bijection means the two sides agree on the partition and
        # disagree only on the numbering (recoverable, an offset); a broken one
        # means a bucket was merged or split (a real algorithm difference).
        pb, bb = mine["wp"][2:], theirs["wp"][2:]
        if fwd.setdefault(pb, bb) != bb:
            fwd["__broken__"] = True
        if rev.setdefault(bb, pb) != pb:
            rev["__broken__"] = True

    if stamped_bad:
        fails.append("%d features where one side stamped and the other did not, e.g. %s"
                     % (len(stamped_bad), stamped_bad[:3]))
    if wf_bad:
        fails.append("family `wf` differs on %d features, e.g. %s"
                     % (len(wf_bad), "; ".join("#%d %s/%s" % t for t in wf_bad[:MAX_SHOWN])))
    if wp_bad:
        fails.append("pattern `wp` differs on %d features, e.g. %s"
                     % (len(wp_bad), "; ".join("#%d %s/%s" % t for t in wp_bad[:MAX_SHOWN])))

    py_combos = sorted(combos)
    br_combos = sorted({r["wp"] for r in brows if r["wp"]})
    if py_combos != br_combos:
        fails.append("combos differ: python-only %s  browser-only %s"
                     % (sorted(set(py_combos) - set(br_combos)),
                        sorted(set(br_combos) - set(py_combos))))

    # ------------------------------------------------------------- report ----
    stamped = sum(1 for r in rows if r)
    bij = "__broken__" not in fwd and "__broken__" not in rev
    print("snapshot               %s" % date)
    print("features               python %d   browser %d" % (len(rows), len(brows)))
    print("assembly               %s" % json.dumps(asm))
    print("stamped                python %d   browser %d"
          % (stamped, sum(1 for r in brows if r["wp"])))
    print("palette buckets        python %d   browser %d" % (len(palette), len(bp)))
    print("(family,bucket) combos python %d   browser %d" % (len(py_combos), len(br_combos)))
    print("bucket labelling       %s" % ("a bijection" if bij else "*** NOT a bijection"))
    print("wp exact matches       %d / %d" % (stamped - len(wp_bad), stamped))
    print("")
    if fails:
        print("*FAIL - %d finding(s)" % len(fails))
        for f in fails[:20]:
            print("   " + f)
        return 1
    print("PASS - the bake reproduces the browser's assembly feature-for-feature, "
          "elects the same %d buckets to the same hex, and stamps all %d buildings "
          "with the same (family, bucket)." % (len(palette), stamped))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
